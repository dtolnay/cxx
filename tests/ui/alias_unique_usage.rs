// This fails because the original bridge does not see any usage of
// UniquePtr<C> and so does not generate the approprite trait. The
// original bridge currently needs to include some usage of UniquePtr<C>.
// TODO: File a Github issue to fix this.

#[cxx::bridge(namespace = here)]
mod here {
    extern "C" {
        type C;
    }
}

// Rustfmt mangles the extern type alias.
// https://github.com/rust-lang/rustfmt/issues/4159
#[rustfmt::skip]
#[cxx::bridge(namespace = there)]
#[cxx::alias_namespace(crate::here = here)]
mod there {
    extern "C" {
        type C = crate::here::C;

        fn c_return_unique_ptr() -> UniquePtr<C>;
    }
}

fn main() {}
