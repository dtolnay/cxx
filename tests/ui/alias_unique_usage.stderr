error[E0277]: the trait bound `here::C: UniquePtrTarget` is not satisfied
  --> $DIR/alias_unique_usage.rs:16:1
   |
16 | #[cxx::bridge(namespace = there)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `UniquePtrTarget` is not implemented for `here::C`
   |
  ::: $WORKSPACE/src/unique_ptr.rs
   |
   |     T: UniquePtrTarget,
   |        --------------- required by this bound in `UniquePtr`
